crates/astria-composer/src/ds/eth_provider.rs

```diff
@@ -0,0 +1,92 @@
+use std::time::Duration;
+
+use async_trait::async_trait;
+use humantime::format_duration;
+use tracing::{debug, info, instrument, warn};
+use color_eyre::eyre::{self, Context};
+use tokio::sync::mpsc as tokio_mpsc;
+
+use ethers::providers::{
+    Middleware, Provider as EthersProvider, ProviderError as EthersProviderError, StreamExt, Ws,
+};
+
+use super::{streaming_client::StreamingClient, RollupChainId, RollupTxExt};
+
+/// A thin wrapper around [`Provider<Ws>`] to add timeouts.
+pub(crate) struct EthProvider {
+    inner: EthersProvider<Ws>,
+}
+
+// TODO(https://github.com/astriaorg/astria/issues/216): add timeouts for
+// `subscribe_full_pendings_txs` (more complex because it's a stream).
+impl EthProvider {
+    pub(crate) async fn connect(url: &str) -> Result<Self, EthersProviderError> {
+        let inner = EthersProvider::connect(url).await?;
+        Ok(Self { inner })
+    }
+
+    /// Wrapper around [`Provider::get_net_version`] with a 1s timeout.
+    async fn get_net_version(&self) -> Result<String, eyre::Error> {
+        tokio::time::timeout(Duration::from_secs(1), self.inner.get_net_version())
+            .await
+            .wrap_err("request timed out")?
+            .wrap_err("RPC returned with error")
+    }
+
+    /// Wait until a connection to eth is established.
+    ///
+    /// This function tries to establish a connection to eth by
+    /// querying its net_version RPC. If it fails, it retries for another `n_retries`
+    /// times with exponential backoff.
+    ///
+    /// # Errors
+    ///
+    /// An error is returned if calling eth failed after `n_retries + 1` times.
+    #[instrument(skip_all, fields(
+        retries.max_number = n_retries,
+        retries.initial_delay = %format_duration(delay),
+        retries.exponential_factor = factor,
+    ))]
+    pub(crate) async fn setup(
```

Comment:
why change the name of this from `wait_for_eth`? also, this should be called before incrementing the `active_providers` count in `searcher::Status` (or before setting it to `true` in the single provider case).